### ISSUE TITLE
Subscriptions without some billing details will not renew after updating to 7.3.0

### ISSUE THREAD

**Issue by csmcneill** [Automattic/woocommerce-payments#8352]
### Describe the bug

If a merchant removes some billing fields (e.g., billing address, country) from the checkout page, the subscriptions would be successfully created and renew without issue prior to 7.3.0.

However, after updating to 7.3.0, these same subscription renewals will fail, and the order notes indicate the following error:

```
Error: You passed an empty string for 'billing_details[address][country]'. We assume empty values are an attempt to unset a parameter; however 'billing_details[address][country]' cannot be unset. You should remove 'billing_details[address][country]' from your request or supply a non-empty value.
```

### To Reproduce

1. Create a simple subscription product
2. Install WooPayments 7.1.0 (7.2.0 may work here, but I didn't test this)
4. Install [Checkout Field Editor](https://woo.com/products/woocommerce-checkout-field-editor/?aff=10486&cid=1131038)
5. Remove all billing fields except ZIP code, email, and first/last name
6. Purchase a subscription via shortcode checkout (CFE is not compatible with the checkout block)
7. Update to WooPayments 7.3.0
8. Process a renewal by running the `woocommerce_scheduled_subscription_payment` hook in the action scheduler.
9. The renewal will fail

### Actual behavior

A renewal fails if there is incomplete billing information provided. However, this only occurs with 7.3.0 active, so it might be related to https://github.com/Automattic/woocommerce-payments/pull/8226

### Screenshots

![CleanShot 2024-03-07 at 18 39 48](https://github.com/Automattic/woocommerce-payments/assets/82999806/c7dc01d3-855c-4c74-b720-33ce80ed4e0e)

### Expected behavior

Since subscriptions renewing with incomplete billing information worked before 7.3.0, maybe it should work in 7.3.0 and beyond — even though [we advise against removing core billing fields](https://arc.net/l/quote/ewtwsbww).

### Additional context

p1709858821739469-slack-C7U3Y3VMY
7855095-zen

**Comment by csmcneill** [comment]
Adding the `impact: high` label as we currently estimate that ~28 sites are impacted per p1709915385055709-slack-CGGCLBN58

**Comment by maxlaf** [comment]
7864552-zen
Just adding cases here for measuring impact.

**Comment by rossviviano** [comment]
7869057-zen

**Comment by claudiulodro** [comment]
I think it would be best to continue supporting as many removed fields as possible. [Research shows that the more fields a checkout has the lower the conversion rate will be](https://baymard.com/blog/checkout-flow-average-form-fields), so removing unnecessary fields (e.g. many parts of the billing address on digital products) provides a meaningful difference to merchants' revenue.

**Comment by csmcneill** [comment]
@gpressutto5 @frosso Could either of you check to see if this is related to #8226 please?

I reverted the changes introduced by this PR on my local and was able to resolve the issue. If I had to wager a guess (and pardon my ignorance here), `get_billing_details_from_order` might be trying to pull billing details on subscription renewals — but when some billing details are not present, the order fails.

**Comment by frosso** [comment]
@csmcneill it does seem to be related. I tested it locally (with and without those fixes) and came to the same conclusion. I think you're on the right track with the issue.

**Comment by zoupkat** [comment]
Hey there @frosso 
Any news on the progress here?
Another affected merchant here 7894412-zen
Thank you!

**Comment by frosso** [comment]
Hey @zoupkat! 👋  The issue is on my team's sprint, someone with availability will self-assign and we'll get a resolution within the next week or two.

**Comment by FangedParakeet** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/8352?planning-poker) with Zenhub @gpressutto5

**Comment by wyter** [comment]
8048923-zen, awaiting fix

**Comment by robb100** [comment]
@gpressutto5 

This issue was marked as closed on 26.3.24 but has returned with the latest version 7.7.

Can you re-open and prioritise the case as billings are failing constantly? 

The 7.4 fix allowed absent address fields to be ignored (which was a fault with 7.3). I've just rolled back to 7.6 and everything still works fine there, so it seems the bug is with 7.7.

Appreciate some urgent attention.

Thanks ✌🏼


#8459 #8352

**Comment by robb100** [comment]
@FangedParakeet any update? Thanks

**Comment by frosso** [comment]
Hi @robb100 ! 👋 

I tested the steps on the issue again by processing the manual renewal with subscriptions created on both WooPayments `7.1.0` and `7.6.0`.

In all cases, the subscription renewed successfully on `7.7.0` and `develop`, both via the `woocommerce_scheduled_subscription_payment` action scheduled action and the "Process renewal" admin action.
<img width="379" alt="Screenshot 2024-06-05 at 4 08 36 PM" src="https://github.com/Automattic/woocommerce-payments/assets/273592/94624e20-0a97-46e5-ace1-b9e7795c9acf">

In all cases, these were the fields at checkout:
<img width="1809" alt="Screenshot 2024-06-05 at 4 07 18 PM" src="https://github.com/Automattic/woocommerce-payments/assets/273592/c3c22dac-ca78-4aca-ad83-0aeedc09780d">

Do you happen to have additional details or reproduction steps that could help us?

**Comment by robb100** [comment]
@frosso Thanks

This is the error I got on all automatic scheduled subscription payments.

<img width="255" alt="Screenshot 2024-06-05 at 15 11 46" src="https://github.com/Automattic/woocommerce-payments/assets/171476295/9a8c7b78-e4b0-4ff1-8fbf-c255794971e6">


Attempting to manually process the renewal also didn't work on 7.7 (I tried this both from the order back end and also from the front end on the site)

Only after rolling back to 7.6 would renewal complete.

Like your settings, all customers have their First and Last names along with Email address.

**Comment by frosso** [comment]
Thank you for the additional details @robb100 !

I tried the following setups:
- Subscription set up with a currency different than the store's default currency (e.g.: if the store's default currency is USD, purchase the subscription with EUR via multi-currency setup)
- Subscription with virtual subscriptions

I am still unable to reproduce the error.
We'll report back after further testing.

**Comment by FangedParakeet** [comment]
I just tested this on a JN, where I was also unable to replicate the failure 😞, with the following plugin versions:

- WooCommerce 8.9.2
- WooCommerce Checkout Field Editor 1.7.14
- WooCommerce Subscriptions 6.3.2

Here are the steps I followed, which I believe follow those in the issue description.

1. Installed WooPayments 7.1.0.
2. Used Checkout Field Editor to remove all billing fields, except for first/last name, email, and postcode.
3. Created a simple subscription product and created an order for this product.
4. Upgraded WooPayments to 7.7.0.
5. Ran the subscription renewal via the Action Scheduler using the `woocommerce_scheduled_subscription_payment` hook.
6. Renewal order created and renewal payment charged successfully, orders updated appropriately.

@robb100, can you provide a little more guidance as to how we might be able to produce this error or share a test site where we might be able instigate and inspect the failure ourselves? 🙏

**Comment by gpressutto5** [comment]
I also tested in JN using different versions, blocks and shortcode checkout, and different ways to renew the subscription, but I could not reproduce the error in version 7.7.0.

When the feature to update billing address before processing payment was reimplemented in #8720, [one of the test cases covered this issue](https://github.com/Automattic/woocommerce-payments/pull/8720#:~:text=Filtering%20billing%20details,renewal%20should%20work).

**Comment by robb100** [comment]
@FangedParakeet @gpressutto5 

I don't have any auto renew payments due on this site for another 2 weeks now. 

(for context I'm not a developer or coder, just someone who manages the back end of my site)

I don't have any further info to give you at this stage or a staging site that I can play around with to re-replicate the issue. All I know is that I couldn't process the renewal payments until the version was rolled back to 7.6.

The Woo Auto-update plugin has re-installed the 7.7 - (is there anyway to override the auto updates? it seems to have no settings for that!)...

Anyway. Thanks for looking in to it - I'll update you when the next autorenewals are processed.

**Comment by frosso** [comment]
@gpressutto5 @FangedParakeet I wanted to raise this internal message, in case you also find it interesting interesting: p1718031304506519-slack-C46H1UYN4

**Comment by frosso** [comment]
@robb100 thank you again for your patience.
We'd like to avoid waiting for possible subscriptions failing on your site.
If you're ok with this, we'd like you to create a support ticket within Woo support. Could you please contact us at [WooCommerce.com > My Account > Support](https://woocommerce.com/my-account/contact-support/), providing your [system status report](https://woocommerce.com/document/woocommerce-self-service-guide/#section-7).
We'd also like to take a closer look with an administrator access on your site, if possible.

Head to _Users > Add New_ and enter "woologin" for the user name and "woologin@woocommerce.com" for the email address. Be sure to set the role to Administrator and click “Add New User” as on the screenshot – ![pslYE8.png](https://github.com/Automattic/woocommerce-payments/assets/273592/42d91f16-23b8-40de-9729-12c6f85e2299)

You won't need to disclose us the password, we'll be able to recover it.

Note that to comply with GDPR rules, you need to disclose that you are sharing access rights to third parties like us in your Terms & Conditions. You can find out more about your responsibilities as a store owner under GDPR here: https://woocommerce.com/posts/the-gdpr-and-you-the-woocommerce-store-owner/

Last, but not least, delete our user account once your interaction with us is done. This helps keep your website safe by limiting the number of users with admin access.

Thank you!

**Comment by robb100** [comment]
@frosso  I've created the Admin user for you. Where do I send the password and 2FA Auth Code/Backup code, and  log in URL to so it reaches you directly.

Thanks again

**Comment by frosso** [comment]
Thank you @robb100 ! You can contact us at [WooCommerce.com > My Account > Support](https://woocommerce.com/my-account/contact-support/), providing your [system status report](https://woocommerce.com/document/woocommerce-self-service-guide/#section-7) and referencing this ticket. This way, all the information will remain confidential and be forwarded to my team.

**Comment by EjayhanFernandes** [comment]
Hi everyone! I was able to assist @robb100 with his issue.

I couldn't replicate the error on WooPayments 7.7.0 on my own test site, so I investigated further and discovered that the problem is related to a [custom snippet](https://wpdavies.dev/remove-checkout-fields-for-downloadable-products-woocommerce/) added to the site that removes the billing address fields for virtual products.

This custom snippet works on WooPayments 7.6.0, allowing users to checkout without billing fields, but it doesn't work on 7.7.0. The failed renewals are due to subscriptions purchased without billing fields on version 7.6.0, which then encounter errors when renewed on version 7.7.0. The renewal orders fail with the following error: 
![aCswu8.png](https://github.com/Automattic/woocommerce-payments/assets/81605775/634d5705-bdfa-488c-a7b2-246df8bdb354)

I think we can close this issue.
cc: @frosso
